src/dune_rules/module_trie.ml

```diff
@@ -103,14 +103,25 @@ let rec to_dyn f t =
     t
 
 let merge x y ~f =
-  let base _path _ = (* TODO *) assert false in
+  let rec base ~path ~f t =
+    Map.foldi t ~init:empty ~f:(fun k v acc ->
+        let path = k :: path in
+        let rev_path = List.rev path in
+        match v with
+        | Leaf leaf ->
+          let leaf = f rev_path leaf in
+          Map.add_exn acc k (Leaf leaf)
+        | Map m -> Map.add_exn acc k (Map (base ~path ~f m)))
+  in
   let rec loop path x y =
     match (x, y) with
     | None, None -> assert false
-    | Some x, None -> base path x
-    | None, Some x -> base path x
+    | Some x, None ->
+      base x ~path ~f:(fun path x -> f path (Some x) None |> Option.value_exn)
```

Comment:
Given that this function is provided by the caller, I don't think we can assume that it never returns `None`.

You could modify its signature if our only uses of it always guarantee there's a value available.

Reply:
Good catch, the latest push handles the option case